README.md
# minimist <sup>[![Version Badge][npm-version-svg]][package-url]</sup>

<!-- [![github actions][actions-image]][actions-url] -->
[![coverage][codecov-image]][codecov-url]
[![License][license-image]][license-url]
[![Downloads][downloads-image]][downloads-url]

[![npm badge][npm-badge-png]][package-url]

parse argument options (esm fork)

This is an ESM fork of [minimist](https://github.com/minimistjs/minimist).

This module is the guts of optimist's argument parser without all the
fanciful decoration. 

# example

``` js
import parseArgs from '@scg82/minimist';
const argv = parseArgs(process.argv.slice(2));
console.log(argv);
```

``` sh
$ node example/parse.js -a beep -b boop
{ _: [], a: 'beep', b: 'boop' }
```

``` sh
$ node example/parse.js -x 3 -y 4 -n5 -abc --beep=boop foo bar baz
{
	_: ['foo', 'bar', 'baz'],
	x: 3,
	y: 4,
	n: 5,
	a: true,
	b: true,
	c: true,
	beep: 'boop'
}
```

# methods

``` js
import parseArgs from '@scg82/minimist';
```

## const argv = parseArgs(args, opts={})

Return an argument object `argv` populated with the array arguments from `args`.

`argv._` contains all the arguments that didn't have an option associated with
them.

Numeric-looking arguments will be returned as numbers unless `opts.string` or
`opts.boolean` is set for that argument name.

Any arguments after `'--'` will not be parsed and will end up in `argv._`.

options can be:

* `opts.string` - a string or array of strings argument names to always treat as
strings
* `opts.boolean` - a boolean, string or array of strings to always treat as
booleans. if `true` will treat all double hyphenated arguments without equal signs
as boolean (e.g. affects `--foo`, not `-f` or `--foo=bar`)
* `opts.alias` - an object mapping string names to strings or arrays of string
argument names to use as aliases
* `opts.default` - an object mapping string argument names to default values
* `opts.stopEarly` - when true, populate `argv._` with everything after the
first non-option
* `opts['--']` - when true, populate `argv._` with everything before the `--`
and `argv['--']` with everything after the `--`.

Here's an example:

  ``` js
  parseArgs('one two three -- four five --six'.split(' '), { '--': true })
  ```
  returns
  ``` js
  {
      _: ['one', 'two', 'three'],
      '--': ['four', 'five', '--six']
  }
  ```

  Note that with `opts['--']` set, parsing for arguments still stops after the
  `--`.

* `opts.unknown` - a function which is invoked with a command line parameter not
defined in the `opts` configuration object. If the function returns `false`, the
unknown option is not added to `argv`.

# install

With [npm](https://npmjs.org) do:

```
npm install @scg82/minimist
```

# license

MIT

[package-url]: https://npmjs.org/package/@scg82/minimist
[npm-version-svg]: https://versionbadg.es/scg82/minimist.svg
[npm-badge-png]: https://nodei.co/npm/@scg82/minimist.png?downloads=true&stars=true
[license-image]: https://img.shields.io/npm/l/@scg82/minimist.svg
[license-url]: LICENSE
[downloads-image]: https://img.shields.io/npm/dm/@scg82/minimist.svg
[downloads-url]: https://npm-stat.com/charts.html?package=@scg82/minimist
[codecov-image]: https://codecov.io/gh/SCG82/minimist/branch/main/graphs/badge.svg
[codecov-url]: https://app.codecov.io/gh/SCG82/minimist/
<!-- [actions-image]: https://img.shields.io/endpoint?url=https://github-actions-badge-u3jn4tfpocch.runkit.sh/minimistjs/minimist -->
[actions-url]: https://github.com/SCG82/minimist/actions


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v2.0.2](https://github.com/SCG82/minimist/compare/v2.0.1...v2.0.2) - 2022-12-31

## v2.0.1 - 2022-12-31

### Merged

- move the `opts['--']` example back where it belongs [`#63`](https://github.com/SCG82/minimist/pull/63)

### Fixed

- [Tests] Remove duplicate test [`#8`](https://github.com/SCG82/minimist/issues/8)
- [Fix] opt.string works with multiple aliases [`#9`](https://github.com/SCG82/minimist/issues/9)

### Commits

- [eslint] fix indentation and whitespace [`5368ca4`](https://github.com/SCG82/minimist/commit/5368ca4147e974138a54cc0dc4cea8f756546b70)
- ESM [`e589c8d`](https://github.com/SCG82/minimist/commit/e589c8dc899f03e4a2a130652a7c97afc528b770)
- [eslint] more cleanup [`62fde7d`](https://github.com/SCG82/minimist/commit/62fde7d935f83417fb046741531a9e2346a36976)
- Merge tag 'v0.2.2' [`980d7ac`](https://github.com/SCG82/minimist/commit/980d7ac61a0b4bd552711251ac107d506b23e41f)
- half of the parse test ported [`3079326`](https://github.com/SCG82/minimist/commit/307932601325087de6cf94188eb798ffc4f3088a)
- dedicated boolean test [`46e448f`](https://github.com/SCG82/minimist/commit/46e448f9f513cfeb2bcc8b688b9b47ba1e515c2b)
- [meta] add `auto-changelog` [`73923d2`](https://github.com/SCG82/minimist/commit/73923d223553fca08b1ba77e3fbc2a492862ae4c)
- [meta] add `auto-changelog` [`0ebf4eb`](https://github.com/SCG82/minimist/commit/0ebf4ebcd5f7787a5524d31a849ef41316b83c3c)
- stripped down code and a passing test from optimist [`7cced88`](https://github.com/SCG82/minimist/commit/7cced88d82e399d1a03ed23eb667f04d3f320d10)
- add support for handling "unknown" options not registered with the parser. [`6f3cc5d`](https://github.com/SCG82/minimist/commit/6f3cc5d4e84524932a6ef2ce3592acc67cdd4383)
- ported parse tests completely over [`9448754`](https://github.com/SCG82/minimist/commit/944875452e0820df6830b1408c26a0f7d3e1db04)
- docs, package.json [`a5bf46a`](https://github.com/SCG82/minimist/commit/a5bf46ac9bb3bd114a9c340276c62c1091e538d5)
- [actions] add reusable workflows [`e115b63`](https://github.com/SCG82/minimist/commit/e115b63fa9d3909f33b00a2db647ff79068388de)
- [actions] add reusable workflows [`d80727d`](https://github.com/SCG82/minimist/commit/d80727df77bfa9e631044d7f16368d8f09242c91)
- reformat package.json [`02ed371`](https://github.com/SCG82/minimist/commit/02ed37115194d3697ff358e8e25e5e66bab1d9f8)
- [eslint] add eslint; rules to enable later are warnings [`48bc06a`](https://github.com/SCG82/minimist/commit/48bc06a1b41f00e9cdf183db34f7a51ba70e98d4)
- [eslint] add eslint; rules to enable later are warnings [`f58745b`](https://github.com/SCG82/minimist/commit/f58745b9bb84348e1be72af7dbba5840c7c13013)
- move more short tests into short.js [`503edb5`](https://github.com/SCG82/minimist/commit/503edb5c41d89c0d40831ee517154fc13b0f18b9)
- dedicated num test [`9bf2d36`](https://github.com/SCG82/minimist/commit/9bf2d36f1d3b8795be90b8f7de0a937f098aa394)
- support all-boolean mode [`450a97f`](https://github.com/SCG82/minimist/commit/450a97f6e2bc85c7a4a13185c19a818d9a5ebe69)
- default bool test was wrong, not the code [`1b9f5db`](https://github.com/SCG82/minimist/commit/1b9f5db4741b49962846081b68518de824992097)
- allow dotted aliases & defaults [`321c33e`](https://github.com/SCG82/minimist/commit/321c33e755485faaeb44eeb1c05d33b2e0a5a7c4)
- check that they type of a value is a boolean, not just that it is currently set to a boolean [`6863198`](https://github.com/SCG82/minimist/commit/6863198e36139830ff1f20ffdceaddd93f2c1db9)
- aliased values treated as strings [`1ab743b`](https://github.com/SCG82/minimist/commit/1ab743bad4484d69f1259bed42f9531de01119de)
- Convert boolean arguments to boolean values [`8f3dc27`](https://github.com/SCG82/minimist/commit/8f3dc27cf833f1d54671b6d0bcb55c2fe19672a9)
- passing long tests ripped out of parse.js [`7972c4a`](https://github.com/SCG82/minimist/commit/7972c4aff1f4803079e1668006658e2a761a0428)
- more failing proto pollution tests [`13c01a5`](https://github.com/SCG82/minimist/commit/13c01a5327736903704984b7f65616b8476850cc)
- return '' if flag is string and empty [`fa63ed4`](https://github.com/SCG82/minimist/commit/fa63ed4651a4ef4eefddce34188e0d98d745a263)
- add failing test - boolean alias array [`0fa3c5b`](https://github.com/SCG82/minimist/commit/0fa3c5b3dd98551ddecf5392831b4c21211743fc)
- upgrade tape, fix type issues from old tape version [`806712d`](https://github.com/SCG82/minimist/commit/806712df91604ed02b8e39aa372b84aea659ee34)
- [Dev Deps] switch from `covert` to `nyc` [`ab03356`](https://github.com/SCG82/minimist/commit/ab033567b9c8b31117cb026dc7f1e592ce455c65)
- Improve '--' handling. [`b11822c`](https://github.com/SCG82/minimist/commit/b11822c09cc9d2460f30384d12afc0b953c037a4)
- [readme] rename and add badges [`5df0fe4`](https://github.com/SCG82/minimist/commit/5df0fe49211bd09a3636f8686a7cb3012c3e98f0)
- [readme] rename and add badges [`236f4a0`](https://github.com/SCG82/minimist/commit/236f4a07e4ebe5ee44f1496ec6974991ab293ffd)
- added stopEarly option [`471c7e4`](https://github.com/SCG82/minimist/commit/471c7e4a7e910fc7ad8f9df850a186daf32c64e9)
- [Dev Deps] switch from `covert` to `nyc` [`a48b128`](https://github.com/SCG82/minimist/commit/a48b128fdb8d427dfb20a15273f83e38d97bef07)
- [Dev Deps] update `covert`, `tape`; remove unnecessary `tap` [`f0fb958`](https://github.com/SCG82/minimist/commit/f0fb958e9a1fe980cdffc436a211b0bda58f621b)
- badges [`84c0370`](https://github.com/SCG82/minimist/commit/84c037063664d42878aace715fe6572ce01b6f3b)
- even more aggressive checks for protocol pollution [`38a4d1c`](https://github.com/SCG82/minimist/commit/38a4d1caead72ef99e824bb420a2528eec03d9ab)

test/num.js
import test from 'tape'
import parse from '../src/index.js'

test('nums', function (t) {
    const argv = parse([
        '-x',
        '1234',
        '-y',
        '5.67',
        '-z',
        '1e7',
        '-w',
        '10f',
        '--hex',
        '0xdeadbeef',
        '789',
    ])
    t.deepEqual(argv, {
        x: 1234,
        y: 5.67,
        z: 1e7,
        w: '10f',
        hex: 0xdeadbeef,
        _: [789],
    })
    t.deepEqual(typeof argv.x, 'number')
    t.deepEqual(typeof argv.y, 'number')
    t.deepEqual(typeof argv.z, 'number')
    t.deepEqual(typeof argv.w, 'string')
    t.deepEqual(typeof argv.hex, 'number')
    t.deepEqual(typeof argv._[0], 'number')
    t.end()
})

test('already a number', function (t) {
    const argv = parse(['-x', 1234, 789])
    t.deepEqual(argv, { x: 1234, _: [789] })
    t.deepEqual(typeof argv.x, 'number')
    t.deepEqual(typeof argv._[0], 'number')
    t.end()
})


test/bool.js
import test from 'tape'
import parse from '../src/index.js'

test('flag boolean default false', function (t) {
    const argv = parse(['moo'], {
        boolean: ['t', 'verbose'],
        default: { verbose: false, t: false },
    })

    t.deepEqual(argv, {
        verbose: false,
        t: false,
        _: ['moo'],
    })

    t.deepEqual(typeof argv.verbose, 'boolean')
    t.deepEqual(typeof argv.t, 'boolean')
    t.end()
})

test('boolean groups', function (t) {
    const argv = parse(['-x', '-z', 'one', 'two', 'three'], {
        boolean: ['x', 'y', 'z'],
    })

    t.deepEqual(argv, {
        x: true,
        y: false,
        z: true,
        _: ['one', 'two', 'three'],
    })

    t.deepEqual(typeof argv.x, 'boolean')
    t.deepEqual(typeof argv.y, 'boolean')
    t.deepEqual(typeof argv.z, 'boolean')
    t.end()
})

test('boolean and alias with chainable api', function (t) {
    const aliased = ['-h', 'derp']
    const regular = ['--herp', 'derp']
    const aliasedArgv = parse(aliased, {
        boolean: 'herp',
        alias: { h: 'herp' },
    })
    const propertyArgv = parse(regular, {
        boolean: 'herp',
        alias: { h: 'herp' },
    })
    const expected = {
        herp: true,
        h: true,
        _: ['derp'],
    }

    t.same(aliasedArgv, expected)
    t.same(propertyArgv, expected)
    t.end()
})

test('boolean and alias with options hash', function (t) {
    const aliased = ['-h', 'derp']
    const regular = ['--herp', 'derp']
    const opts = {
        alias: { h: 'herp' },
        boolean: 'herp',
    }
    const aliasedArgv = parse(aliased, opts)
    const propertyArgv = parse(regular, opts)
    const expected = {
        herp: true,
        h: true,
        _: ['derp'],
    }
    t.same(aliasedArgv, expected)
    t.same(propertyArgv, expected)
    t.end()
})

test('boolean and alias array with options hash', function (t) {
    const aliased = ['-h', 'derp']
    const regular = ['--herp', 'derp']
    const alt = ['--harp', 'derp']
    const opts = {
        alias: { h: ['herp', 'harp'] },
        boolean: 'h',
    }
    const aliasedArgv = parse(aliased, opts)
    const propertyArgv = parse(regular, opts)
    const altPropertyArgv = parse(alt, opts)
    const expected = {
        harp: true,
        herp: true,
        h: true,
        _: ['derp'],
    }
    t.same(aliasedArgv, expected)
    t.same(propertyArgv, expected)
    t.same(altPropertyArgv, expected)
    t.end()
})

test('boolean and alias using explicit true', function (t) {
    const aliased = ['-h', 'true']
    const regular = ['--herp', 'true']
    const opts = {
        alias: { h: 'herp' },
        boolean: 'h',
    }
    const aliasedArgv = parse(aliased, opts)
    const propertyArgv = parse(regular, opts)
    const expected = {
        herp: true,
        h: true,
        _: [],
    }

    t.same(aliasedArgv, expected)
    t.same(propertyArgv, expected)
    t.end()
})

// regression, see https://github.com/substack/node-optimist/issues/71
test('boolean and --x=true', function (t) {
    let parsed = parse(['--boool', '--other=true'], {
        boolean: 'boool',
    })

    t.same(parsed.boool, true)
    t.same(parsed.other, 'true')

    parsed = parse(['--boool', '--other=false'], {
        boolean: 'boool',
    })

    t.same(parsed.boool, true)
    t.same(parsed.other, 'false')
    t.end()
})

test('boolean --boool=true', function (t) {
    const parsed = parse(['--boool=true'], {
        default: {
            boool: false,
        },
        boolean: ['boool'],
    })

    t.same(parsed.boool, true)
    t.end()
})

test('boolean --boool=false', function (t) {
    const parsed = parse(['--boool=false'], {
        default: {
            boool: true,
        },
        boolean: ['boool'],
    })

    t.same(parsed.boool, false)
    t.end()
})

test('boolean using something similar to true', function (t) {
    const opts = { boolean: 'h' }
    const result = parse(['-h', 'true.txt'], opts)
    const expected = {
        h: true,
        _: ['true.txt'],
    }

    t.same(result, expected)
    t.end()
})


test/dash.js
import test from 'tape'
import parse from '../src/index.js'

test('-', function (t) {
    t.plan(5)
    t.deepEqual(parse(['-n', '-']), { n: '-', _: [] })
    t.deepEqual(parse(['-']), { _: ['-'] })
    t.deepEqual(parse(['-f-']), { f: '-', _: [] })
    t.deepEqual(parse(['-b', '-'], { boolean: 'b' }), { b: true, _: ['-'] })
    t.deepEqual(parse(['-s', '-'], { string: 's' }), { s: '-', _: [] })
})

test('-a -- b', function (t) {
    t.plan(2)
    t.deepEqual(parse(['-a', '--', 'b']), { a: true, _: ['b'] })
    t.deepEqual(parse(['--a', '--', 'b']), { a: true, _: ['b'] })
})

test('move arguments after the -- into their own `--` array', function (t) {
    t.plan(1)
    t.deepEqual(
        parse(['--name', 'John', 'before', '--', 'after'], { '--': true }),
        { name: 'John', _: ['before'], '--': ['after'] },
    )
})


index.js
function hasKey(obj, keys) {
	let o = obj;
	keys.slice(0, -1).forEach(function (key) {
		o = o[key] || {};
	});

	const key = keys[keys.length - 1];
	return key in o;
}

function isNumber(x) {
	if (typeof x === 'number') { return true; }
	if ((/^0x[0-9a-f]+$/i).test(x)) { return true; }
	return (/^[-+]?(?:\d+(?:\.\d*)?|\.\d+)(e[-+]?\d+)?$/).test(x);
}

function isConstructorOrProto(obj, key) {
	return (key === 'constructor' && typeof obj[key] === 'function') || key === '__proto__';
}

export default function minimist(args, opts) {
	if (!opts) { opts = {}; }

	const flags = {
		bools: {},
		strings: {},
		unknownFn: null,
	};

	if (typeof opts.unknown === 'function') {
		flags.unknownFn = opts.unknown;
	}

	if (typeof opts.boolean === 'boolean' && opts.boolean) {
		flags.allBools = true;
	} else {
		[].concat(opts.boolean).filter(Boolean).forEach(function (key) {
			flags.bools[key] = true;
		});
	}

	const aliases = {};

	function aliasIsBoolean(key) {
		return aliases[key].some(function (x) {
			return flags.bools[x];
		});
	}

	Object.keys(opts.alias || {}).forEach(function (key) {
		aliases[key] = [].concat(opts.alias[key]);
		aliases[key].forEach(function (x) {
			aliases[x] = [key].concat(aliases[key].filter(function (y) {
				return x !== y;
			}));
		});
	});

	[].concat(opts.string).filter(Boolean).forEach(function (key) {
		flags.strings[key] = true;
		if (aliases[key]) {
			[].concat(aliases[key]).forEach(function (k) {
				flags.strings[k] = true;
			});
		}
	});

	const defaults = opts.default || {};

	const argv = { _: [] };

	function argDefined(key, arg) {
		return (flags.allBools && (/^--[^=]+$/).test(arg))
			|| flags.strings[key]
			|| flags.bools[key]
			|| aliases[key];
	}

	function setKey(obj, keys, value) {
		let o = obj;
		for (let i = 0; i < keys.length - 1; i++) {
			const key = keys[i];
			if (isConstructorOrProto(o, key)) { return; }
			if (o[key] === undefined) { o[key] = {}; }
			if (
				o[key] === Object.prototype
				|| o[key] === Number.prototype
				|| o[key] === String.prototype
			) {
				o[key] = {};
			}
			if (o[key] === Array.prototype) { o[key] = []; }
			o = o[key];
		}

		const lastKey = keys[keys.length - 1];
		if (isConstructorOrProto(o, lastKey)) { return; }
		if (
			o === Object.prototype
			|| o === Number.prototype
			|| o === String.prototype
		) {
			o = {};
		}
		if (o === Array.prototype) { o = []; }
		if (o[lastKey] === undefined || flags.bools[lastKey] || typeof o[lastKey] === 'boolean') {
			o[lastKey] = value;
		} else if (Array.isArray(o[lastKey])) {
			o[lastKey].push(value);
		} else {
			o[lastKey] = [o[lastKey], value];
		}
	}

	function setArg(key, val, arg) {
		if (arg && flags.unknownFn && !argDefined(key, arg)) {
			if (flags.unknownFn(arg) === false) { return; }
		}

		const value = !flags.strings[key] && isNumber(val)
			? Number(val)
			: val;
		setKey(argv, key.split('.'), value);

		(aliases[key] || []).forEach(function (x) {
			setKey(argv, x.split('.'), value);
		});
	}

	Object.keys(flags.bools).forEach(function (key) {
		setArg(key, defaults[key] === undefined ? false : defaults[key]);
	});

	let notFlags = [];

	if (args.indexOf('--') !== -1) {
		notFlags = args.slice(args.indexOf('--') + 1);
		args = args.slice(0, args.indexOf('--'));
	}

	for (let i = 0; i < args.length; i++) {
		const arg = args[i];
		let key;
		let next;

		if ((/^--.+=/).test(arg)) {
			// Using [\s\S] instead of . because js doesn't support the
			// 'dotall' regex modifier. See:
			// http://stackoverflow.com/a/1068308/13216
			const m = arg.match(/^--([^=]+)=([\s\S]*)$/);
			key = m[1];
			let value = m[2];
			if (flags.bools[key]) {
				value = value !== 'false';
			}
			setArg(key, value, arg);
		} else if ((/^--no-.+/).test(arg)) {
			key = arg.match(/^--no-(.+)/)[1];
			setArg(key, false, arg);
		} else if ((/^--.+/).test(arg)) {
			key = arg.match(/^--(.+)/)[1];
			next = args[i + 1];
			if (
				next !== undefined
				&& !(/^-/).test(next)
				&& !flags.bools[key]
				&& !flags.allBools
				&& (aliases[key] ? !aliasIsBoolean(key) : true)
			) {
				setArg(key, next, arg);
				i += 1;
			} else if ((/^(true|false)$/).test(next)) {
				setArg(key, next === 'true', arg);
				i += 1;
			} else {
				setArg(key, flags.strings[key] ? '' : true, arg);
			}
		} else if ((/^-[^-]+/).test(arg)) {
			const letters = arg.slice(1, -1).split('');

			let broken = false;
			for (let j = 0; j < letters.length; j++) {
				next = arg.slice(j + 2);

				if (next === '-') {
					setArg(letters[j], next, arg);
					continue;
				}

				if ((/[A-Za-z]/).test(letters[j]) && (/[=]/).test(next)) {
					setArg(letters[j], next.split('=')[1], arg);
					broken = true;
					break;
				}

				if (
					(/[A-Za-z]/).test(letters[j])
					&& (/-?\d+(\.\d*)?(e-?\d+)?$/).test(next)
				) {
					setArg(letters[j], next, arg);
					broken = true;
					break;
				}

				if (letters[j + 1] && letters[j + 1].match(/\W/)) {
					setArg(letters[j], arg.slice(j + 2), arg);
					broken = true;
					break;
				} else {
					setArg(letters[j], flags.strings[letters[j]] ? '' : true, arg);
				}
			}

			key = arg.slice(-1)[0];
			if (!broken && key !== '-') {
				if (
					args[i + 1]
					&& !(/^(-|--)[^-]/).test(args[i + 1])
					&& !flags.bools[key]
					&& (aliases[key] ? !aliasIsBoolean(key) : true)
				) {
					setArg(key, args[i + 1], arg);
					i += 1;
				} else if (args[i + 1] && (/^(true|false)$/).test(args[i + 1])) {
					setArg(key, args[i + 1] === 'true', arg);
					i += 1;
				} else {
					setArg(key, flags.strings[key] ? '' : true, arg);
				}
			}
		} else {
			if (!flags.unknownFn || flags.unknownFn(arg) !== false) {
				argv._.push(flags.strings._ || !isNumber(arg) ? arg : Number(arg));
			}
			if (opts.stopEarly) {
				argv._.push.apply(argv._, args.slice(i + 1));
				break;
			}
		}
	}

	Object.keys(defaults).forEach(function (k) {
		if (!hasKey(argv, k.split('.'))) {
			setKey(argv, k.split('.'), defaults[k]);

			(aliases[k] || []).forEach(function (x) {
				setKey(argv, x.split('.'), defaults[k]);
			});
		}
	});

	if (opts['--']) {
		argv['--'] = notFlags.slice();
	} else {
		notFlags.forEach(function (k) {
			argv._.push(k);
		});
	}

	return argv;
}


src/types.ts
export interface Flags {
    unknownFn?: (arg: string) => boolean
    allBools?: boolean
    bools: Record<string, boolean>
    strings: Record<string, boolean>
}

export interface Opts {
    /**
     * A string or array of strings argument names to always treat as strings
     */
    string?: string | string[] | undefined

    /**
     * A boolean, string or array of strings to always treat as booleans. If true will treat
     * all double hyphenated arguments without equals signs as boolean (e.g. affects `--foo`, not `-f` or `--foo=bar`)
     */
    boolean?: boolean | string | string[] | undefined

    /**
     * An object mapping string names to strings or arrays of string argument names to use as aliases
     */
    alias?: { [key: string]: string | string[] } | undefined

    /**
     * An object mapping string argument names to default values
     */
    default?: { [key: string]: any } | undefined

    /**
     * When true, populate argv._ with everything after the first non-option
     */
    stopEarly?: boolean | undefined

    /**
     * A function which is invoked with a command line parameter not defined in the opts
     * configuration object. If the function returns false, the unknown option is not added to argv
     */
    unknown?: ((arg: string) => boolean) | undefined

    /**
     * When true, populate argv._ with everything before the -- and argv['--'] with everything after the --.
     * Note that with -- set, parsing for arguments still stops after the `--`.
     */
    '--'?: boolean | undefined
}

export interface ParsedArgs {
    [arg: string]: any

    /**
     * If opts['--'] is true, populated with everything after the --
     */
    '--'?: string[] | undefined

    /**
     * Contains all the arguments that didn't have an option associated with them
     */
    _: (string | number | boolean)[]
}


src/utilities.ts
import type { Flags, ParsedArgs } from './types'

export function aliasIsBoolean(
    alias: string[],
    bools: Record<string, boolean>,
) {
    return alias.some((x) => bools[x])
}

export function argDefined(
    key: string | number,
    arg: string,
    flags: Flags,
    aliases: Record<string, string[]>,
) {
    return (
        (flags.allBools && /^--[^=]+$/.test(arg)) ||
        flags.strings[key] ||
        flags.bools[key] ||
        aliases[key]
    )
}

export function hasKey(obj: ParsedArgs, keys: string[]) {
    let o = obj
    keys.slice(0, -1).forEach((key) => {
        o = o[key] ?? {}
    })
    const key = keys[keys.length - 1]
    return key in o
}

export function isConstructorOrProto(
    obj: object | ((...args: any[]) => any) | string | number | undefined,
    key: unknown,
) {
    if (typeof obj === 'undefined') {
        return false
    }
    return (
        (key === 'constructor' && typeof obj[key] === 'function') ||
        key === '__proto__'
    )
}

export function isNumber(x: unknown): x is number {
    if (typeof x === 'number') {
        return true
    }
    if (/^0x[0-9a-f]+$/i.test(String(x))) {
        return true
    }
    return /^[-+]?(?:\d+(?:\.\d*)?|\.\d+)(e[-+]?\d+)?$/.test(String(x))
}


src/index.ts
import process from 'node:process'
import type { Flags, Opts, ParsedArgs } from './types.js'
import {
    aliasIsBoolean,
    argDefined,
    hasKey,
    isConstructorOrProto,
    isNumber,
} from './utilities.js'

/**
 * Return an argument object populated with the array arguments from args
 *
 * @param [args] An optional argument array @default process.argv.slice(2)
 * @param [opts] An optional options object to customize the parsing
 */
export function minimist(args?: string[], opts?: Opts): ParsedArgs

/**
 * Return an argument object populated with the array arguments from args. Strongly-typed
 * to be the intersect of type T with minimist.ParsedArgs.
 *
 * `T` The type that will be intersected with minimist.ParsedArgs to represent the argument object
 *
 * @param [args] An optional argument array @default process.argv.slice(2)
 * @param [opts] An optional options object to customize the parsing
 */
export function minimist<T>(args?: string[], opts?: Opts): T & ParsedArgs

/**
 * Return an argument object populated with the array arguments from args. Strongly-typed
 * to be the the type T which should extend minimist.ParsedArgs
 *
 * `T` The type that extends minimist.ParsedArgs and represents the argument object
 *
 * @param [args] An optional argument array @default process.argv.slice(2)
 * @param [opts] An optional options object to customize the parsing
 */
export function minimist<T extends ParsedArgs>(
    args: string[] = process.argv.slice(2),
    opts: Opts = {},
): T {
    const flags: Flags = {
        bools: {},
        strings: {},
        unknownFn: undefined,
    }

    if (typeof opts.unknown === 'function') {
        flags.unknownFn = opts.unknown
    }

    if (typeof opts.boolean === 'boolean' && opts.boolean) {
        flags.allBools = true
    } else {
        Array.prototype
            .concat([], opts.boolean)
            .filter(Boolean)
            .forEach((key) => {
                flags.bools[key] = true
            })
    }

    const aliases: Record<string, string[]> = {}

    if (typeof opts.alias === 'object') {
        Object.entries(opts.alias).forEach(([key, val]) => {
            aliases[key] = Array.prototype.concat([], val)
            aliases[key].forEach((x) => {
                aliases[x] = [key].concat(aliases[key].filter((y) => x !== y))
            })
        })
    }

    Array.prototype
        .concat([], opts.string)
        .filter(Boolean)
        .forEach((key) => {
            flags.strings[key] = true
            if (aliases[key]) {
                Array.prototype.concat([], aliases[key]).forEach((k) => {
                    flags.strings[k] = true
                })
            }
        })

    const defaults = opts.default || {}
    const argv = { _: [] } as unknown as T

    // function argDefined(key: string | number, arg: string) {
    //     return (
    //         (flags.allBools && /^--[^=]+$/.test(arg)) ||
    //         flags.strings[key] ||
    //         flags.bools[key] ||
    //         aliases[key]
    //     )
    // }

    function setKey(obj: ParsedArgs, keys: (string | number)[], value: any) {
        let o: any = obj
        for (let i = 0; i < keys.length - 1; i++) {
            const key = keys[i]
            if (isConstructorOrProto(o, key)) {
                return
            }
            if (o[key] === undefined) {
                o[key] = {}
            }
            if (
                o[key] === Object.prototype ||
                o[key] === Number.prototype ||
                o[key] === String.prototype
            ) {
                o[key] = {}
            }
            if (o[key] === Array.prototype) {
                o[key] = []
            }
            o = o[key]
        }

        const lastKey = keys[keys.length - 1]
        if (isConstructorOrProto(o, lastKey)) {
            return
        }
        if (
            o === Object.prototype ||
            o === Number.prototype ||
            o === String.prototype
        ) {
            o = {}
        }
        if (o === Array.prototype) {
            o = []
        }
        if (
            o[lastKey] === undefined ||
            flags.bools[lastKey] ||
            typeof o[lastKey] === 'boolean'
        ) {
            o[lastKey] = value
        } else if (Array.isArray(o[lastKey])) {
            o[lastKey].push(value)
        } else {
            o[lastKey] = [o[lastKey], value]
        }
    }

    function setArg(
        key: string | undefined,
        val: string | boolean | undefined,
        arg?: string,
    ) {
        if (key === undefined || val === undefined) {
            return
        }
        if (arg && flags.unknownFn && !argDefined(key, arg, flags, aliases)) {
            if (flags.unknownFn(arg) === false) {
                return
            }
        }

        const value = !flags.strings[key] && isNumber(val) ? Number(val) : val
        setKey(argv, key.split('.'), value)
        ;(aliases[key] || []).forEach((x) => {
            setKey(argv, x.split('.'), value)
        })
    }

    Object.keys(flags.bools).forEach((key) => {
        setArg(key, defaults[key] === undefined ? false : defaults[key])
    })

    let notFlags: string[] = []

    if (args.indexOf('--') !== -1) {
        notFlags = args.slice(args.indexOf('--') + 1)
        args = args.slice(0, args.indexOf('--'))
    }

    for (let i = 0; i < args.length; i++) {
        const arg = args[i]
        let match: RegExpMatchArray | null
        let key: string | undefined
        let next: string | undefined

        if (/^--.+=/.test(arg)) {
            /**
             * Using `[\s\S]` instead of `.` because js doesn't support the 'dotall' regex modifier.
             * @see http://stackoverflow.com/a/1068308/13216
             */
            const m = arg.match(/^--([^=]+)=([\s\S]*)$/)
            if (m) {
                key = m[1]
                let value: string | boolean | undefined = m[2]
                if (key !== undefined && flags.bools?.[key]) {
                    value = value !== 'false'
                }
                setArg(key, value, arg)
            }
        } else if ((match = String(arg).match(/^--no-(.+)/))) {
            key = match[1]
            setArg(key, false, arg)
        } else if ((match = String(arg).match(/^--(.+)/))) {
            key = match[1]
            next = args[i + 1]
            if (
                // key !== undefined &&
                next !== undefined &&
                !/^-/.test(next) &&
                !flags.bools[key] &&
                !flags.allBools &&
                (aliases[key]
                    ? !aliasIsBoolean(aliases[key], flags.bools)
                    : true)
            ) {
                setArg(key, next, arg)
                i += 1
            } else if (/^(true|false)$/.test(next)) {
                setArg(key, next === 'true', arg)
                i += 1
            } else if (key) {
                setArg(key, flags.strings[key] ? '' : true, arg)
            }
        } else if (/^-[^-]+/.test(String(arg))) {
            const letters = arg.slice(1, -1).split('')

            let broken = false
            for (let j = 0; j < letters.length; j++) {
                next = arg.slice(j + 2)

                if (next === '-') {
                    setArg(letters[j], next, arg)
                    continue
                }

                if (/[A-Za-z]/.test(letters[j]) && /[=]/.test(next)) {
                    setArg(letters[j], next.split('=')[1], arg)
                    broken = true
                    break
                }

                if (
                    /[A-Za-z]/.test(letters[j]) &&
                    /-?\d+(\.\d*)?(e-?\d+)?$/.test(next)
                ) {
                    setArg(letters[j], next, arg)
                    broken = true
                    break
                }

                if (letters[j + 1] && letters[j + 1].match(/\W/)) {
                    setArg(letters[j], arg.slice(j + 2), arg)
                    broken = true
                    break
                }

                setArg(letters[j], flags.strings[letters[j]] ? '' : true, arg)
            }

            key = arg.slice(-1)[0]
            if (!broken && key !== '-') {
                if (
                    args[i + 1] &&
                    !/^(-|--)[^-]/.test(args[i + 1]) &&
                    !flags.bools[key] &&
                    (aliases[key]
                        ? !aliasIsBoolean(aliases[key], flags.bools)
                        : true)
                ) {
                    setArg(key, args[i + 1], arg)
                    i += 1
                } else if (args[i + 1] && /^(true|false)$/.test(args[i + 1])) {
                    setArg(key, args[i + 1] === 'true', arg)
                    i += 1
                } else {
                    setArg(key, flags.strings[key] ? '' : true, arg)
                }
            }
        } else {
            if (!flags.unknownFn || flags.unknownFn(arg) !== false) {
                argv._.push(
                    flags.strings._ || !isNumber(arg) ? arg : Number(arg),
                )
            }
            if (opts.stopEarly) {
                argv._.push(...args.slice(i + 1))
                break
            }
        }
    }

    Object.keys(defaults).forEach((k) => {
        if (!hasKey(argv, k.split('.'))) {
            setKey(argv, k.split('.'), defaults[k])
            ;(aliases[k] || []).forEach((x) => {
                setKey(argv, x.split('.'), defaults[k])
            })
        }
    })

    if (opts['--']) {
        argv['--'] = notFlags.slice()
    } else {
        notFlags.forEach(function (k) {
            argv._.push(k)
        })
    }

    return argv
}

export default minimist
